src/hardware/sblaster.cpp

```diff
@@ -1995,25 +1996,47 @@ static bool write_buffer_at_capacity()
 	return (sb.dma.left > sb.dma.min);
 }
 
+// Sound Blaster DSP status byte
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+// Ref: http://archive.gamedev.net/archive/reference/articles/article443.html
+//
+// Read from 02x0Ch (DSP Write Buffer Status): Bit 7, when 0, indicates that the
+// DSP is ready to receive data through the DSP Write Data or Command port
+// (02x0Ch write).
+//
+// Read from 02x0Eh (DSP Data Available Status): Bit 7, when 1, indicates that
+// the DSP has pending data to be read through the DSP Read Data port (02x0Ah
+// read).
+//
+union BufferStatus {
+	uint8_t data = 0;
+	bit_view<7, 1> has_data;
+	// when 1, the buffer has data and is ready to send
+	// when 0, the buffer doesn't have data and is ready to receive
+};
+
 static uint8_t read_sb(io_port_t port, io_width_t)
 {
 	switch (port - sb.hw.base) {
 	case MIXER_INDEX: return sb.mixer.index;
 	case MIXER_DATA: return CTMIXER_Read();
 	case DSP_READ_DATA: return DSP_ReadData();
-	case DSP_READ_STATUS:
+	case DSP_READ_STATUS: {
 		// TODO See for high speed dma :)
 		if (sb.irq.pending_8bit) {
 			sb.irq.pending_8bit = false;
 			PIC_DeActivateIRQ(sb.hw.irq);
 		}
-		if (sb.dsp.out.used) {
-			return 0xff;
-		} else {
-			return 0x7f;
-		}
+		BufferStatus read_status = {};
+		read_status.has_data = (sb.dsp.out.used != 0);
+		return read_status.data;
+	}
```

Comment:
are we officially OK with type punning in cases like this?

Reply:
(it's how all the `bit_view` unions work.. if we're not OK with then, there's a lot we're not OK with :-)

`bit_view` is just a vehicle to operate the data `.data` member:

```c++
// all-purpose assignment
template <class rhs_type>
constexpr bit_view& Assign(const rhs_type& rhs_value) noexcept
{
	if constexpr (std::is_integral_v<rhs_type>) {
		check_rhs(rhs_value);
	}
	const auto outer = data & ~mask::shifted;
	const auto inner = (rhs_value & mask::unshifted) << view_index;

	data = static_cast<data_type>(outer | inner);
	return *this;
}
```

So the write goes straight to `.data` and then we read `.data` (in this case).

But same with reading bits; they all boil down to reads on `.data`:

```c++
// read the view's value
constexpr operator data_type() const noexcept
{
return (data & mask::shifted) >> view_index;
}
```

Reply:
@kklobe - to eliminate the union, maybe `<variant>` can be used under the hood? I don't know enough about it though.

Reply:
well - `memcpy` is one way around it 😁. C++20 will give us `bit_cast`, so maybe we just let it ride for now.

Reply:
(I should also say that the `bit_view` unit tests pass GCC and Clang's undefined behaviour sanitizers on Linux and macOS; so I think this is pretty much solid until we want to swap it out :slightly_smiling_face: !)

Reply:
> are we officially OK with type punning in cases like this?

Yes 😎 (short and sweet)

...what @kcgen said, keep the end goal in mind. Bit views are super useful and increase the clarity of the code a great deal. That's what we should optimise for, whatever it takes, so I'm 100% fine.

Reply:
Yes, but it's UB, so I'm clarifying we are ok with that.

Reply:
I don't think it's UB because under the hood, we always write to the same union member (the data member) as we read from.

Provided those are always done back to back, it's not UB.

@kklobe, are you getting any compiler or sanitizer warnings from it?

Reply:
Hmm, interesting, It's still a union, but given the slight of hand, I wonder if it would fall outside of https://en.cppreference.com/w/cpp/language/union#Explanation. I don't get compiler warnings, but none of the other non-bitview instances in the codebase throw warnings either, so it clearly must not be a very big deal.

Reply:
Could it be that it exclusives uses `uint8_t`'s? It never moves from one type to another (ie: write a 32-bit to a union `float` but then read those bits back as a union-punned `int32_t`… that's definitely UB, I think).

It's used throughout the code now, and passes Coverity and PVS Studio too (as well as sanitizer builds).

Undefined behaviour is unwanted behaviour; absolutely - I want Staging passing all the sanitizers. And certainly - if there's a better way to get the job done, let's do it!

Reply:
Sounds like a plan; thanks @kklobe.